src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al

```diff
@@ -0,0 +1,518 @@
+
+        exit(CompletionAnswerTxt);
+    end;
+
+    local procedure GenerateNoSeriesUsingToolResult(var AzureOpenAI: Codeunit "Azure OpenAI"; var AOAIChatMessages: Codeunit "AOAI Chat Messages"; var AOAIChatCompletionParams: Codeunit "AOAI Chat Completion Params"; var AOAIOperationResponse: Codeunit "AOAI Operation Response"): Text
+    var
+        AOAIFunctionResponse: Codeunit "AOAI Function Response";
+        NoSeriesCopToolsImpl: Codeunit "No. Series Cop. Tools Impl.";
+        ToolResponse: Text;
+        ToolResponses: Dictionary of [Text, Integer]; // tool response can be a list of strings, as the response can be too long and exceed the token limit. In this case each string would be a separate message, each of them should be called separately. The integer is the number of tables used in the prompt, so we can test if the LLM answer covers all tables
+        FinalResults: List of [Text]; // The final response will be the concatenation of all the LLM responses (final results).
+        ExpectedNoSeriesCount: Integer;
+        Progress: Dialog;
+        GeneratingNoSeriesForLbl: Label 'Generating number series %1', Comment = '%1 = No. Series';
+    begin
+        // remove the tools from the chat messages, as they are not needed anymore
+        AOAIChatMessages.ClearTools();
+
+        AOAIFunctionResponse := AOAIOperationResponse.GetFunctionResponse();
+        if not AOAIFunctionResponse.IsSuccess() then
+            Error(AOAIFunctionResponse.GetError());
+
+        ToolResponses := AOAIFunctionResponse.GetResult();
+
+        if ToolResponses.Count = 0 then
+            Error(NotAbleToGenerateNumberSeriesTryToRephraseErr);
+
+        AOAIChatMessages.SetPrimarySystemMessage(GetNoSeriesGenerationSystemPrompt());
+        AOAIChatCompletionParams.SetJsonMode(true);
+
+        foreach ToolResponse in ToolResponses.Keys() do begin
+            Progress.Open(StrSubstNo(GeneratingNoSeriesForLbl, NoSeriesCopToolsImpl.ExtractAreaWithPrefix(ToolResponse)));
+            // adding function response to messages
+            AOAIChatMessages.AddToolMessage(AOAIFunctionResponse.GetFunctionId(), AOAIFunctionResponse.GetFunctionName(), ToolResponse);
+
+            // call the API again to get the final response from the model
+            ToolResponses.Get(ToolResponse, ExpectedNoSeriesCount);
+            if not GenerateAndReviewToolCompletionWithRetry(AzureOpenAI, AOAIChatMessages, AOAIChatCompletionParams, ExpectedNoSeriesCount) then
+                Error(GetLastErrorText());
+
+            FinalResults.Add(AOAIChatMessages.GetLastMessage());
+
+            AOAIChatMessages.DeleteMessage(AOAIChatMessages.GetHistory().Count); // remove the last message, as it is not needed anymore
+            AOAIChatMessages.DeleteMessage(AOAIChatMessages.GetHistory().Count); // remove the tools message, as it is not needed anymore
+
+            Sleep(1000); // sleep for 1000ms, as the model can be called only limited number of times per second
```

Comment:
is 1000ms really the right number? do we have any plan to finetune this a little more?

Reply:
Removed